File: 2025/03/03.py
#!/usr/bin/env python

# Copyrigt (C) 2025 Joelle Maslak
# All Rights Reserved - See License

import sys
from typing import List


def main():
    batt_list: List[List[int]] = []

    # Read Battery strings
    txt = ""
    for line in sys.stdin.readlines():
        batt_list.append([int(x) for x in line.strip()])

    # Set counter variables
    part1: int = 0
    part2: int = 0

    # Do the work
    for batts in batt_list:
        part1 += int(largest_joltage(batts, 2))
        part2 += int(largest_joltage(batts, 12))

    print(f"Solution for part 1: {part1}")
    print(f"Solution for part 2: {part2}")


def largest_joltage(batts: List[int], sz: int) -> str:
    assert sz >= 0

    if sz == 0:
        return ""

    # These are all the possible first batteries, by value
    possibles = batts[slice(0,len(batts)-sz)]
    maxval = max(possibles)  # The actual biggest possible first battery

    for i in range(0, len(batts)):
        val = batts[i]
        if val == maxval:
            # We have the biggest starting battery size, so now we recurse.
            remainder = batts[slice(i+1, len(batts))]
            return f"{val}{largest_joltage(remainder, sz-1)}"

    assert 0 == 1 # Shouldn't get here.
    return ""


if __name__ == "__main__":
    main()

File: 2025/03/03.pl
#!/usr/bin/env perl

#
# Copyright (C) 2025 Joelle Maslak
# All Rights Reserved - See License
#

use lib '.';

use JTM::Boilerplate 'script';
use List::Util qw(max uniqstr);

MAIN: {
    my @strings;
    while ( <<>> ) {
        chomp;
        push @strings, [split "", $_];
    }

    my $part1 = 0;
    my $part2 = 0;

    for my $string (@strings) {
        $part1 += largest($string, 2);
        $part2 += largest($string, 12);
    }

    say("Part 1: $part1");
    say("Part 2: $part2");
}

sub largest($string, $sz) {
    if ($sz == 0) {
        return "";
    }
    my $len = scalar(@$string);
    my $lastidx = $len-$sz;

    my @possibles = @$string[0..$lastidx];
    my $max = max(@possibles);

    my $ret = 0;
    my %seen;  # Used as speedup.
    for (my $i=0; $i<=$lastidx; $i++) {
        my $val = $string->[$i];
        if ($val != $max) { next }
        my (@remainder) = @$string[($i+1)..($len-1)];
        if (scalar(@remainder)) {
            if (exists $seen{$val . $remainder[0]}) { next }
            $seen{$val . $remainder[0]} = 1;
        }
        my $x = "$val" . largest(\@remainder, $sz-1);
        if ($x > $ret) {
            $ret = $x;
        }
    }

    return $ret
}
